node/src/ios/OpenT2TClient/ObjCppUtils.h

```diff
@@ -0,0 +1,81 @@
+
+void ExceptionToNSError(std::exception_ptr ex, NSError** outError)
+{
+    if (ex != nullptr)
+    {
+        // TODO: Use other domains/error codes for some exceptions.
+        NSString* domain = @"OpenT2T";
+        NSInteger code = 1;
+        NSString* description = nil;
+
+        try
+        {
+            std::rethrow_exception(ex);
+        }
+        catch (const std::out_of_range& stdex)
+        {
+            description = [NSString stringWithUTF8String: stdex.what()];
+        }
+        catch (const std::length_error& stdex)
+        {
+            description = [NSString stringWithUTF8String: stdex.what()];
+        }
+        catch (const std::invalid_argument& stdex)
+        {
+            description = [NSString stringWithUTF8String: stdex.what()];
+        }
+        catch (const std::bad_cast& stdex)
+        {
+            description = [NSString stringWithUTF8String: stdex.what()];
+        }
+        catch (const std::logic_error& stdex)
+        {
+            description = [NSString stringWithUTF8String: stdex.what()];
+        }
+        catch (const std::runtime_error& stdex)
+        {
+            description = [NSString stringWithUTF8String: stdex.what()];
+        }
+        catch (const std::exception& stdex)
+        {
+            description = [NSString stringWithUTF8String: stdex.what()];
+        }
+        catch (...)
+        {
+        }
+
+        NSDictionary* userInfo;
+        if (description != nil)
+        {
+            OpenT2T::LogTrace("ExceptionToNSError(\"%s\")", description.UTF8String);
+            userInfo = @{ NSLocalizedDescriptionKey : description };
+        }
+        else
+        {
+            OpenT2T::LogTrace("ExceptionToNSError(\"\")");
+            userInfo = @{};
+        }
+
+        if (outError != nil)
+        {
+            *outError = [[NSError alloc] initWithDomain: domain code: code userInfo: userInfo];
+        }
+    }
+}
+
+bool ValidateArgumentNotNull(const char* methodName, const char* argName, NSObject* arg, NSError** outError)
+{
+    if (arg == nil)
+    {
+        std::string message = std::string() +
```

Comment:
NSString is UTF16. So that would cause conversion from UTF16 to UTF8 (for `LogError` and `invalid_argument`) and back to UTF16 again for `NSError`. This way at least there's only one conversion.

---
In reply to: [71141463](https://github.com/openT2T/client/pull/1#discussion_r71141463) [](ancestors = 71141463)